coralproject/talk
# [CORL-2754] Optimize filling the archive queue

## What does this PR do?

Make the fill archiving queue job operation walk a cursor to find stories to archive instead of running a large convoluted query against the entire stories collection.

Instead of running an expensive query, we can cheaply iterate the cursor from Mongo and do the checks ourselves. This prevents a slow query bogging down Mongo every interval during the scheduled archiving times and uses our readily available compute in our job pods.

## These changes will impact:


## What changes to the GraphQL/Database Schema does this PR introduce?

None

## Does this PR introduce any new environment variables or feature flags? 

No

## If any indexes were added, were they added to `INDEXES.md`?

No new indexes.

## How do I test this PR?

- start up Coral
- create a story
- edit that story's createdAt date so that it's older than 6 months ago
    - for instance, if `createdAt` was `2023-03-13T16:28:38.090+00:00`, make it something like `2020-03-13T16:28:38.090+00:00`
- stop Coral
- set the following in your `.env` file:
    - `MONGODB_ARCHIVE_URI` to a valid mongo instance URI
    - `DISABLE_JOB_PROCESSORS=false`
    - `ENABLE_AUTO_ARCHIVING=true`
    - `AUTO_ARCHIVING_INTERVAL=0,30 * * * * *`
        - this will make auto archiving run every 30 seconds
- start up coral with `npm run start:development:withJobs`
    - I added this option because the default `npm run start:development` does not run jobs
- watch your Coral logs, you'll see it schedule to fill the auto archiving queue and then a `retrieveStoriesToBeArchived` will run and it should have a `count`
    - if no other stories were archivable, the count should be 1
    - if you have a lot of stories in your collection, it might be larger since it will also try and find other stories to add into the archiving queue
- wait for the logs to say it has finished archiving the story you modified
- check your story, see that it is archived
 
## How do we deploy this PR?

No special considerations.
